### ISSUE TITLE
AnnotationInteractionDelegate incompatible with detecting clicks on the map

### ISSUE THREAD

**Issue by akirmse** [mapbox/mapbox-maps-ios#761]
In https://github.com/mapbox/mapbox-maps-ios/issues/598, I noted that it was impossible to have more than one AnnotationManager reporting taps, because each installed its own global gesture recognizer.  While that has been fixed, it is now impossible for an app to detect clicks on the map that are *not* on any annotation, because those are not reported via the AnnotationInteractionDelegate.  Prior to the previous bug being fixed, they were reported via a call to didDetectTappedAnnotations with an empty annotation list, but this appears to no longer be the case.

The workaround is for the app to install its own UITapGestureRecognizer and do all collision detection itself manually, making AnnotationInteractionDelegate superfluous.

Possible fixes:

- Reinstate the calls to didDetectTappedAnnotations with an empty annotation list for clicks on the map.  That is a little awkward because the delegate would be called once for each annotation manager delegate.  The app could disambiguate by ignoring calls from all but one annotation manager (say, the PointAnnotationManager).

- Add a new protocol or callback on MapboxMap or MapView that is called specifically when a click occurs on the map, but not on an annotation.

**Comment by akirmse** [comment]
No, all taps, with any associated annotations.  I am replicating the Google Maps behavior, where clicking on an annotation brings up a detail popup window, and clicking on the map itself dismisses it.

**Comment by adamkaump** [comment]
+1 to this request. Being able to detect when a user taps off of an annotation is very important.

**Comment by akirmse** [comment]
I disagree with marking this a feature rather than a bug, BTW.  I cannot use AnnotationInteractionDelegate in my app, because I also need to detect clicks on the map.

**Comment by PaulC-ManyHats** [comment]
I verified that this is indeed a bug between having both UITapGestureRecognizer AND AnnotationInteractionDelegate on the same map.

To reproduce this, open the "SelectAnnotationExample" class and add a UITapGestureRecognizer to the mapView (in viewDidLoad).

`    override func viewDidLoad() {
        super.viewDidLoad()

       ....
        let tapGesture = UITapGestureRecognizer(target: self, action: #selector(handleMapTap))
        mapView.addGestureRecognizer(tapGesture)
    }
    
    @objc func handleMapTap(sender: UITapGestureRecognizer) {
    }
`

After adding the UITapGestureRecognizer the example no longer works.

**Comment by carlodenardin** [comment]
There is a solution to this problem?

Because I'm not able to handle the problem related to the `AnnotationInteractionDelegate`. The user should interact with the map to close a popup. But if I place a `UITapGestureRecognizer` the user can not interact with the `AnnotationPoint`.

**Comment by akirmse** [comment]
My workaround is to install my own ```UITapGestureRecognizer``` and then replicate part of the internals of the ```PointAnnotationManager```.  The most important part is the call to ```map.queryRenderedFeatures```:

```
    // Handle a click on the given screen point.  If it's a click on a marker, generate that event.
    // Otherwise, continue by passing click to handlePolylineClick().
    private func handleMapClick(point: CGPoint) {
        let options = RenderedQueryOptions(layerIds: [self.pointAnnotationManager.layerId], filter: nil)
        map.queryRenderedFeatures(at: point, options: options) { (result) in
            var pointsClicked = [PointAnnotation]()
            switch result {
            case .success(let queriedFeatures):
                let annotationIds: [String] = queriedFeatures.compactMap {
                    guard case let .string(featureId) = $0.feature.identifier else {
                        return nil
                    }
                    return featureId
                }
               pointsClicked = self.pointAnnotationManager.annotations.filter { annotationIds.contains($0.id) }
                
            case .failure:
                // Not sure what this means, but treat it as no point clicked
                ()
            }
            
            // Skip unclickable markers, then sort by z-index
            let clickableMarkers = pointsClicked.filter { point in
                MapboxMarker.isAnnotationClickable(point)
            }.sorted(by: { e1, e2 in (e1.symbolSortKey ?? 0) > (e2.symbolSortKey ?? 0) })
            
            if let annotation = clickableMarkers.first {
                // Click on info window?
                if annotation.id == (self.currentInfoWindow as? MapboxMarker)?.mapboxMarker.id {
                    if let marker = self.markerWithInfoWindow {
                        self.mapDelegate.onInfoWindowClicked(map: self, marker: marker)
                    }
                    return
                }

                let marker = MapboxMarker(marker: annotation)

                // Override default behavior of showing an info window?
                if self.mapDelegate.onMarkerClick(map: self, marker: marker) == true {
                    return
                }
                
                // Close any existing info window and open new one
                self.removeCurrentInfoWindow()
                self.showInfoWindow(marker)
                
                // Move camera to center on annotation
                self.mapDelegate.onCameraWillMove(map: self, isGesture: true)
                self.cameraController.toLatLng(latlng: marker.position, animate: true)
            } else {
                // Not a click on a marker: see if it's a click on a polyline or just the map
                self.handlePolylineClick(point: point)
            }
        }
    }
```

(```MapboxMarker``` is my own wrapper class.)

Then I do pretty much the same thing with my ```PolylineAnnotationManager``` to see if it's a click on a polyline.  If not, it's a click on the map itself.

**Comment by tamanois** [comment]
I also encountered that problem and solved it like this. Instead of adding a  new UITapgesturerecognizer use an existing one, the singleTapGestureRecognizer (Which forward gesture to annotationInteractionDelegate) .

`      mapView!.gestures.singleTapGestureRecognizer.addTarget(self, action: #selector(handleTap(gestureRecognizer:)))
`    
    and the handling function

```
    @objc func handleTap(gestureRecognizer: UITapGestureRecognizer) {
       // Handle tap gesture
     }
```

**Comment by maclne** [comment]
> I also encountered that problem and solved it like this. Instead of adding a new UITapgesturerecognizer use an existing one, the singleTapGestureRecognizer (Which forward gesture to annotationInteractionDelegate) .
> 
> ` mapView!.gestures.singleTapGestureRecognizer.addTarget(self, action: #selector(handleTap(gestureRecognizer:)))` and the handling function
> 
> ```
>     @objc func handleTap(gestureRecognizer: UITapGestureRecognizer) {
>        // Handle tap gesture
>      }
> ```
> 
> Where mapview is the mapview object

Adding a target to the gestureRecognizer, however if you want to tell if it was a tap on an annotation vs a tap on the map, I'm not sure that is possible with this approach?

**Comment by trunghvbk** [comment]
Hi everyone, was this issue resolved? I need this feature also to handle both tapping on the map and the annotations.

**Comment by persidskiy** [comment]
Hi everyone, the issue is resolved in [11.0.0-beta.4](https://github.com/mapbox/mapbox-maps-ios/releases/tag/v11.0.0-beta.4). You can now assign the gesture handlers to the annotations and the map. The annotations will handle the gesture in the rendered order. If no annotation handles the gesture, the map handler will.

```swift
let annotationManager = mapView.annotations.makePolygonAnnotationManager()
var annotation = PolygonAnnotation(...)
annotation.tapHandler = { context in
  print("tapped annotation")
  return true // handled, don't propagate
}
annotationManager.annotations = [annotation]
mapView.gestures.onMapTap.observe { context in
    print("tapped map off-annotation")
}.store(in: &cancelables)
```

Or, in SwiftUI:
```swift
Map {
  PointAnnotation(...)
     .onTapGesture { print("annotation tap") }
}
.onMapTapGesture { _ in print("off-annotation tap") }
```


You can find more information in the [migration guide](https://docs.mapbox.com/ios/maps/api/11.0.0-beta.4-docc/documentation/mapboxmaps/migrate-to-v11#24-Map-Content-Gesture-System) and [gestures guide](https://docs.mapbox.com/ios/maps/api/11.0.0-beta.4-docc/documentation/mapboxmaps/map-content-gestures-user-guide).
